### PR TITLE
ElastiCache ReplicationGroup auth token generation and connection details

### LINKED ISSUE #1102: [Elasticache] - Allow option to autogenerate password for auth
### What problem are you facing?

We are looking at migrating from the [community AWS provider](https://doc.crds.dev/github.com/crossplane/provider-aws/cache.aws.crossplane.io/ReplicationGroup/v1beta1@v0.46.0) to the upbound ones. One significant difference we see is that this provider strictly requires a secret being passed in for auth, whilst we can see the use for this it would also be good to have the ability to autogenerate the password as we currently do and send the result out to `writeConnectionSecretToRef` in a similar way to what happens when creating dynamodb tables using this provider. 

The reason we have this requirement is our infrastructure is entirely gitops based including the manifests used in crossplane so any situation where we have to provide passwords means entering them into git which isn't ideal

### How could Official AWS Provider help solve your problem?
Without this migrating to the upbound provider is quite difficult if not potentially impossible (investigations still ongoing)

### PR DESCRIPTION
**Pull request by chlunde**
- Add option autoGenerateAuthToken for ElastiCache ReplicationGroup
- Add port/endpoint connection details for ElastiCache ReplicationGroup

### Description of your changes

Fixes #1102

I have:

- [x] Read and followed Crossplane's [contribution process].
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Uptest run: https://github.com/crossplane-contrib/provider-upjet-aws/actions/runs/9397299773

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by haarchri** [comment]
@chlunde i am interested too in this PR - do you need help to finish ?

**Comment by chlunde** [comment]
@haarchri I pushed again now. I changed the example region to us-east-1 and added a VPC to get a complete example. The existing region has only two az-s per account so it's not good for examples.

I'm a bit confused as I don't remember getting an error about needing a VPC to use auth tokens the last time I tested, but now I do. This is why I added a VPC to the example.

**Comment by chlunde** [comment]
```
kind: Secret
apiVersion: v1
data:
  attribute.auth_token: VDN4S2dVaFNsVFllMVFScmpmUFNGYTB5aUl4
  port: NjM3OQ==
  primary_endpoint_address: ...==
  reader_endpoint_address: ...=
```

File: apis/elasticache/v1beta2/zz_replicationgroup_types.go [hunk 1]
```diff
@@ -396,13 +397,20 @@ type ReplicationGroupParameters struct {
...
 
 	// Strategy to use when updating the auth_token. Valid values are SET, ROTATE, and DELETE. Defaults to ROTATE.
 	// +kubebuilder:validation:Optional
 	AuthTokenUpdateStrategy *string `json:"authTokenUpdateStrategy,omitempty" tf:"auth_token_update_strategy,omitempty"`
 
+	// Password used to access a password protected server. Can be specified only if transit_encryption_enabled = true.
+	// If true, the auth token will be auto-generated and stored in the Secret referenced by the authTokenSecretRef field.
+	// +upjet:crd:field:TFTag=-
+	// +kubebuilder:validation:Optional
+	AutoGenerateAuthToken *bool `json:"autoGenerateAuthToken,omitempty" tf:"-"`
```
Comment (ulucinar) [inline apis/elasticache/v1beta2/zz_replicationgroup_types.go:412]: nit: In theory, we could also support auto token generation for the `spec.initProvider` API for completeness. But we do not need to do this in this PR as I don't expect it to be a real use case.

**Review by ulucinar — COMMENTED** [review]
Hi @chlunde,
Thank you for taking care of this. Left some comments for discussion. The most critical one is the error handling comment.

File: examples/elasticache/v1beta2/replicationgroup.yaml [hunk 2]
```diff
@@ -14,13 +68,25 @@ metadata:
 spec:
   forProvider:
     automaticFailoverEnabled: true
+    atRestEncryptionEnabled: true
+    autoGenerateAuthToken: true
+    authTokenSecretRef:
+      name: redis-auth-token
+      namespace: crossplane-system
```
Comment (ulucinar) [inline examples/elasticache/v1beta2/replicationgroup.yaml:75]: The namespace `crossplane-system` is not available by default. Let's change this to `upbound-system`.

File: examples/elasticache/v1beta2/replicationgroup.yaml [hunk 3]
```diff
@@ -14,13 +68,25 @@ metadata:
...
     preferredCacheClusterAzs:
-      - us-west-1a
-      - us-west-1b
-    region: us-west-1
+      - us-east-1a
+      - us-east-1b
+    region: us-east-1
+  writeConnectionSecretToRef:
+    name: redis-conn
+    namespace: crossplane-system
```
Comment (ulucinar) [inline examples/elasticache/v1beta2/replicationgroup.yaml:92]: The namespace `crossplane-system` is not available by default. Let's change this to `upbound-system`.

**Comment by turkenf** [comment]
Hi @chlunde, thank you for your effort in this PR. 

The `v1.6.0` release is planned for tomorrow, June 6, we want to include this PR in tomorrow's release. If you do not have time to address review comments, please let us know.

File: apis/elasticache/v1beta2/zz_replicationgroup_types.go [hunk 4]
```diff
@@ -396,13 +397,20 @@ type ReplicationGroupParameters struct {
...
 	// Password used to access a password protected server. Can be specified only if transit_encryption_enabled = true.
+	// If you set autoGenerateAuthToken to true, the Secret referenced here will be created or updated with generated auth token if it does not already contain one.
 	// +kubebuilder:validation:Optional
 	AuthTokenSecretRef *v1.SecretKeySelector `json:"authTokenSecretRef,omitempty" tf:"-"`
 
 	// Strategy to use when updating the auth_token. Valid values are SET, ROTATE, and DELETE. Defaults to ROTATE.
 	// +kubebuilder:validation:Optional
 	AuthTokenUpdateStrategy *string `json:"authTokenUpdateStrategy,omitempty" tf:"auth_token_update_strategy,omitempty"`
 
+	// Password used to access a password protected server. Can be specified only if transit_encryption_enabled = true.
```
[the remark this answers is not in the document]
Comment (chlunde) [reply apis/elasticache/v1beta2/zz_replicationgroup_types.go:408]: This is the same for similar field for RDS Instance, at the moment I don't know where the code that prepends this string is 🤔

File: apis/elasticache/v1beta2/zz_replicationgroup_types.go (hunk 1 above)
Comment (chlunde) [reply apis/elasticache/v1beta2/zz_replicationgroup_types.go:412]: 👍  I think we should do the same every place common.PasswordGenerator is used then (in another PR)

**Comment by chlunde** [comment]
@turkenf @ulucinar Thanks for the review! I think the two remaining comments apply to common.PasswordGenerator in general?

**Comment by turkenf** [comment]
/test-examples="examples/elasticache/v1beta2/replicationgroup.yaml"

**Review by turkenf — APPROVED** [review]
Thanks a lot @chlunde, LGTM.

**Pull request merged**